torch/_ops.py

```diff
@@ -309,7 +310,27 @@ def dispatch(self, dispatch_key, *args, **kwargs):
                 return handler(mode, *args, **kwargs)
 
         functionality_key = torch._C._to_functionality_key(dispatch_key)  # type: ignore[attr-defined]
-        if functionality_key in mode_stack_per_key():
+        if functionality_key == torch._C.DispatchKey.PreDispatch:
+            from torch.utils._python_dispatch import _pop_mode_temporarily
+
+            if (
+                _len_torch_dispatch_stack_pre_dispatch() > 0
+                and not torch._C._dispatch_tls_is_dispatch_key_excluded(
+                    DispatchKey.Python
+                )
+            ):
+                curr_mode = _get_current_dispatch_mode_pre_dispatch()
+                assert (
+                    curr_mode is not None
+                ), "Illegal invocation of dispatch on torch._C.DispatchKey.PreDispatch without a mode."
+                assert (
+                    type(curr_mode) in self.python_key_mode_table
+                ), f"Current active mode {curr_mode} not registered"
+                handler = self.python_key_mode_table[type(curr_mode)]
+                with _pop_mode_temporarily(functionality_key) as mode:
+                    return handler(mode, *args, **kwargs)
+
+        elif functionality_key in mode_stack_per_key():
```

Comment:
We might want to kill `mode_stack_per_key()` as a followup, if we're only ever using it for PreDispatch and we now special-case pre-dispatch above anyway

Reply:
Yup! makes sense